### ISSUE TITLE
Add pytorch Datasets loader to datasets

### ISSUE THREAD

**Issue by davidslater** [twosixlabs/armory#455]
This requires using PyTorch DataLoader to read directly from the TFDS datasets. There may be some issues making this performant, as PyTorch uses very different methods of loading the data.
See this thread: https://discuss.pytorch.org/t/read-dataset-from-tfrecord-format/16409

Most of what I have found focuses on reading TFRecord files (which are protobufs), e.g., https://github.com/vahidk/tfrecord
but does not leverage the larger dataset_info.json and related txt/json files that would be necessary for parsing an entire dataset automatically.

**Comment by davidslater** [comment]
It is possible that we would need to have mirrored PyTorch Dataset files for each dataset, in order to handle parsing the bytes properly, but I would like to avoid this if possible.

**Comment by seanpmorgan** [comment]
So dug into this and think we have 3 options going forward:

1. [Vahidk/tfrecord package](https://github.com/vahidk/tfrecord). So the biggest issue with using this method is that it's very restrictive in what format it expects data to be in. For example it has functionality to load arrays from byte strings but it assumes that they're ready to be [consumed from buffer by numpy](https://github.com/vahidk/tfrecord/blob/b5d0bddf0cbe14e6aea9a1585d186e36a847248a/tfrecord/reader.py#L166). Often we need to decode an image from compressed format, and this would require forking the repo (making it more extensible) and then using opencv or some other library to decode image compression. Also as mentioned you need to parse the dataset_info.json etc. as you mentioned, but that is a very do-able task.

2. [NVIDIA DALI](https://github.com/NVIDIA/DALI), is a really nice tool aimed at speeding up pipelines and being able to leverage the GPU for pre-processing. It has support for reading TFRecord files and returning PyTorch DataLoaders (among a number of other output and input types). It can leverage GPU to do data augmentation, normalization etc so it has a lot of benefits for TF and PyTorch. If we went this route we could consider using DALI for all frameworks we support. The problem is that it requires GPU + Driver to work and so it's not capable of CPU only evaluations. Perhaps this is straying too far from the core functionality of Armory, but it may be something we need to consider if adversarial training and other high computation defenses have the most success. 

3. Load data from NumPy and save the DataLoader to disk. We'll want some checks to verify the data is the same after conversion, as well as a mean to automate this. It'll ~2x our disk usage on S3 which isn't a major concern.

Here's just a quick example of using DALI on resisc45 tfrecord:
```
import os
import nvidia.dali.tfrecord as tfrec
from nvidia.dali.pipeline import Pipeline
import nvidia.dali.ops as ops
import nvidia.dali.types as types
from nvidia.dali.plugin.pytorch import DALIGenericIterator


class CommonPipeline(Pipeline):
    def __init__(self, batch_size, num_threads, device_id):
        super(CommonPipeline, self).__init__(batch_size, num_threads, device_id)

        self.decode = ops.ImageDecoder(device="cpu", output_type=types.RGB)

    def base_define_graph(self, inputs):
        images = self.decode(inputs)
        return images


class TFRecordPipeline(CommonPipeline):
    def __init__(self, batch_size, num_threads, device_id):
        super(TFRecordPipeline, self).__init__(batch_size, num_threads, device_id)
        self.input = ops.TFRecordReader(
            path=tfrecord,
            index_path=tfrecord_idx,
            features={"image": tfrec.FixedLenFeature((), tfrec.string, ""),},
        )

    def define_graph(self):
        inputs = self.input(name="Reader")
        images = inputs["image"]
        return self.base_define_graph(images)


if __name__ == "__main__":
    tfrecord = os.path.join("resisc/resisc45_split-test.tfrecord-00000-of-00001")
    tfrecord_idx = "tfds-pytorch.index"

    pipes = [
        TFRecordPipeline(batch_size=1, num_threads=2, device_id=0)
    ]
    pipes[0].build()
    dali_iter = DALIGenericIterator(pipes, ["image"], pipes[0].epoch_size("Reader"))
    data = next(dali_iter)
    print(data[0]['image'].shape) # torch.Size([1, 256, 256, 3])
```

**Comment by davidslater** [comment]
CMU/Bosch request.
